docs: Concurrent read/write update

mark-boer opened the pull request:
This PR closes issue  #1811

I've updated the threadpool example with the code from https://rasterio.groups.io/g/main/topic/28593754. Unfortunately, when I changed the `ThreadPool` into a `ProcessPool` it resulted in a deadlock. I've added the old example to also show an option using `ProcessPoolExecutor`.

I searched around for documentation on the `shared` option in `rasterio.open` and in GDAL. Would it be interesting to mention this option in this section of the docs? I couldn't find very much information on this flag.

sgillies commented:
Thanks @mark-boer ! I'll take a look. I'm also going to update #1811 with some thoughts about concurrency examples.

sgillies commented on docs/topics/concurrency.rst:
```diff
@@ -43,7 +43,10 @@ a GIL-releasing raster processing function.
+Here is the program in examples/thread_pool_executor.py. It is set up in such
+a way that at most 1 thread is reading and at most 1 thread is writing at the
+same time. Processing is not protected by a lock and can be done by multiple
+threads simultaneously.
```
:+1:

sgillies commented on docs/topics/concurrency.rst:
```diff
@@ -70,40 +75,37 @@ Here is the program in examples/thread_pool_executor.py.
+                # from multiple threads, we use a lock to protect the
+                # DatasetReader/Writer
+                read_lock = multiprocessing.Lock()
+                write_lock = multiprocessing.Lock()
```
Should these be `threading.Lock`?

mark-boer replied:
You are totally right, I was testing around with the `ProcessPoolExecutor`, after that I forgot to change that back.

sgillies commented on docs/topics/concurrency.rst:
```diff
@@ -70,40 +75,37 @@ Here is the program in examples/thread_pool_executor.py.
+                def process(window):
+                    with read_lock:
+                        src_array = src.read(window=window)
+                    result = compute(src_array)
```
Add some white space around line 99?

I think the important thing to get across is that if `compute()` releases the GIL, it can truly execute concurrently in different worker threads. If `compute()` uses, for example, numpy operations (see https://scipy-cookbook.readthedocs.io/items/ParallelProgramming.html#Threads) it can execute partially concurrently  in different threads.

mark-boer replied:
That's true, but even if Threads do not release the GIL, you can get a decent speedup, because 1 thread can be reading, 1 thread computing and 1 thread writing. But this of course depends largely on where the bottleneck is.

I'll read it through one more time to see if I can add some more info on the GIL. I find it quite hard to find information on numpy and multithreading. It seems that in some cases numpy already uses multithreading internally and multithreading it manually is not really needed.

sgillies replied:
That's a good point in your first paragraph.

sgillies commented on docs/topics/concurrency.rst:
```diff
@@ -114,22 +116,40 @@ spread over multiple cores using the ``ThreadPoolExecutor`` from Python 3's
+``ProcessPoolExecutor``, the DatasetReader/Writer cannot be shared by multiple
+processes, which means that each process needs to open the file seperately,
+or you can do all the reading and writing from the main thread, as shown in 
+this next example. This is much less efficient memory wise, however.
```
:+1:

sgillies reviewed (changes requested):
@mark-boer thanks for your work on this! I have just a small suggestion and a question about where the locks come from.

mark-boer commented:
I added a bit more background in the intro and I gave the note about dask a more prominent place in this docs and added a link to numba. I can easily undo this if you think it is a bit much.

sgillies reviewed (approved):
Thanks @mark-boer !

mark-boer commented:
cheers! and thanks for the review @sgillies  😄
